main.vala

```diff
@@ -639,291 +514,1289 @@ class Vls.Server {
         }
+
+        assert (best != null);
+        var sr = best.source_reference;
+        var from = (long)Server.get_string_pos (file.content, sr.begin.line-1, sr.begin.column-1);
+        var to = (long)Server.get_string_pos (file.content, sr.end.line-1, sr.end.column);
+        string contents = file.content [from:to];
+        debug ("Got best node: %s @ %s = %s", best.type_name, sr.to_string(), contents);
+
+        return (!) best;
+    }
+
+    Vala.Scope get_current_scope (Vala.CodeNode code_node) {
+        Vala.Scope? best = null;
+
+        for (Vala.CodeNode? node = code_node; node != null; node = node.parent_node) {
+            if (node is Vala.Symbol) {
+                var sym = (Vala.Symbol) node;
+                best = sym.scope;
+                break;
+            }
+        }
+
+        assert (best != null);
+
+        return (!) best;
     }
 
     void textDocumentDefinition (Jsonrpc.Server self, Jsonrpc.Client client, string method, Variant id, Variant @params) {
         var p = parse_variant <LanguageServer.TextDocumentPositionParams> (@params);
-        debug ("get definition in %s at %u,%u", p.textDocument.uri,
-            p.position.line, p.position.character);
-        var sourcefile = ctx.get_source_file (p.textDocument.uri);
+        TextDocument? sourcefile = lookup_source_file (p.textDocument.uri);
         if (sourcefile == null) {
-            debug ("unknown file %s", p.textDocument.uri);
+            debug (@"[$method] file `$(p.textDocument.uri)' not found");
+            reply_null (id, client, method);
+            return;
+        }
+        var file = sourcefile.file;
+
+        wait_for_context_update (id, request_cancelled => {
+            if (request_cancelled) {
+                reply_null (id, client, method);
+                return;
+            }
+
+            var fs = new FindSymbol (file, p.position.to_libvala ());
+
+            if (fs.result.size == 0) {
+                try {
+                    client.reply (id, new Variant.maybe (VariantType.VARIANT, null));
+                } catch (Error e) {
+                    debug("[textDocument/definition] failed to reply to client: %s", e.message);
+                }
+                return;
+            }
+
+            Vala.CodeNode? best = get_best (fs, sourcefile);
+
+            if (best is Vala.Expression && !(best is Vala.Literal)) {
+                var b = (Vala.Expression)best;
+                debug ("best (%p) is a Expression (symbol_reference = %p)", best, b.symbol_reference);
+                if (b.symbol_reference != null && b.symbol_reference.source_reference != null) {
+                    best = b.symbol_reference;
+                    debug ("best is now the symbol_referenece => %p (%s)", best, best.to_string ());
+                }
+            } else if (best is Vala.DataType) {
+                var dt = best as Vala.DataType;
+                if (dt.type_symbol != null)
+                    best = dt.type_symbol;
```

Comment:
in `DataType`, `type_symbol` refers to the same thing as `symbol`

Reply:
Do you think it could change in the future?

Reply:
no